src/Microsoft.FeatureManagement/FeatureFilters/TimeWindowFilter.cs

```diff
@@ -56,7 +68,74 @@ public Task<bool> EvaluateAsync(FeatureFilterEvaluationContext context)
                 return Task.FromResult(false);
             }
 
-            return Task.FromResult((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value));
+            //
+            // Hit the first occurrence of the time window
+            if ((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value))
+            {
+                return Task.FromResult(true);
+            }
+
+            if (settings.Recurrence != null)
+            {
+                //
+                // The reference of the object will be used for hash key.
+                // If there is no pre-bounded settings attached to the context, there will be no cached filter settings and each call will have a unique settings object.
+                // In this case, the cache for recurrence settings won't work.
+                if (context.Settings != null)
+                {
+                    DateTimeOffset cachedTime = _recurrenceCache.GetOrAdd(
+                        settings,
+                        (_) =>
+                        {
+                            if (RecurrenceEvaluator.TryFindPrevAndNextOccurrences(now, settings, out DateTimeOffset prevOccurrence, out DateTimeOffset _))
+                            {
+                                return prevOccurrence;
+                            }
+
+                            //
+                            // There is no previous occurrence within the reccurrence range.
+                            return DateTimeOffset.MaxValue;
+                        });
+
+                    if (now < cachedTime)
```

Comment:
If there's no previous occurrence, you set the cached value to MaxValue right above. Wouldn't that mean always returning false from that point forward?

Reply:
I covered this scenario in the test suite (Line 1581 in Tests.FeatureManagement/RecurrenceEvaluation.cs)


If there's no previous occurrence (haven't reached start yet), `TryFindPrevAndNextOccurrences` will return true and prevOccurrence will be set to MinValue.

``` C#
public static bool TryFindPrevAndNextOccurrences(DateTimeOffset time, TimeWindowFilterSettings settings, out DateTimeOffset prevOccurrence, out DateTimeOffset nextOccurrence)
{
    prevOccurrence = DateTimeOffset.MinValue;

    nextOccurrence = DateTimeOffset.MaxValue;

    if (time < settings.Start.Value)
    {
        //
        // The time is before the first occurrence.
        nextOccurrence = settings.Start.Value;

        return true;
    }
```
But this implementation seems wired. I will take the suggestion to modify the `TryFindPrevAndNextOccurrences` method to `void CalculateSurroundingOccurrences`